framework/src/convergence/DefaultMultiAppFixedPointConvergence.C

```diff
@@ -0,0 +1,214 @@
+    _accept_max_it(getSharedExecutionerParam<bool>("accept_on_max_fixed_point_iteration")),
+    _fixed_point_rel_tol(getSharedExecutionerParam<Real>("fixed_point_rel_tol")),
+    _fixed_point_abs_tol(getSharedExecutionerParam<Real>("fixed_point_abs_tol")),
+    _custom_pp_rel_tol(getSharedExecutionerParam<Real>("custom_rel_tol")),
+    _custom_pp_abs_tol(getSharedExecutionerParam<Real>("custom_abs_tol")),
+    _fixed_point_custom_pp(isParamValid("custom_pp") ? &getPostprocessorValue("custom_pp")
+                                                     : nullptr),
+    _pp_old(0.0),
+    _pp_new(std::numeric_limits<Real>::max()),
+    _pp_scaling(1.0),
+    _fe_problem(*getCheckedPointerParam<FEProblemBase *>("_fe_problem_base")),
+    _fp_solve(getMooseApp().getExecutioner()->fixedPointSolve())
+{
+  if (_min_fixed_point_its > _max_fixed_point_its)
+    paramError("fixed_point_min_its",
+               "The minimum number of fixed point iterations may not exceed the maximum.");
+  if (!_has_fixed_point_norm && parameters.isParamSetByUser("fixed_point_rel_tol"))
+    paramWarning(
+        "disable_fixed_point_residual_norm_check",
+        "fixed_point_rel_tol will be ignored because the fixed point residual check is disabled.");
+  if (!_has_fixed_point_norm && parameters.isParamSetByUser("fixed_point_abs_tol"))
+    paramWarning(
+        "disable_fixed_point_residual_norm_check",
+        "fixed_point_abs_tol will be ignored because the fixed point residual check is disabled.");
+  if (!_has_fixed_point_norm && parameters.isParamSetByUser("fixed_point_force_norms"))
+    paramWarning("disable_fixed_point_residual_norm_check",
+                 "fixed_point_force_norms will be ignored because the fixed point residual check "
+                 "is disabled.");
+
+  if (dynamic_cast<SteffensenSolve *>(&_fp_solve) && _fe_problem.hasMultiApps())
+  {
+    // Steffensen method uses half-steps
+    if (!parameters.isParamSetByAddParam("fixed_point_min_its"))
+      _min_fixed_point_its *= 2;
+    _max_fixed_point_its *= 2;
+  }
+}
+
+void
+DefaultMultiAppFixedPointConvergence::initialize()
+{
+  DefaultConvergenceBase::initialize();
+
+  _fixed_point_timestep_begin_norm.clear();
+  _fixed_point_timestep_end_norm.clear();
+  _fixed_point_timestep_begin_norm.resize(_max_fixed_point_its);
+  _fixed_point_timestep_end_norm.resize(_max_fixed_point_its);
+
+  _pp_history.str("");
+
+  // compute residual norm before iteration loop
+  if (_has_fixed_point_norm)
+  {
+    _fixed_point_initial_norm = _fe_problem.computeResidualL2Norm();
+    _console << COLOR_MAGENTA << "Initial fixed point residual norm: " << COLOR_DEFAULT;
+    if (_fixed_point_initial_norm == std::numeric_limits<Real>::max())
+      _console << " MAX ";
+    else
+      _console << std::scientific << _fixed_point_initial_norm;
+    _console << COLOR_DEFAULT << "\n" << std::endl;
+  }
+}
+
+void
+DefaultMultiAppFixedPointConvergence::preExecute()
+{
+  DefaultConvergenceBase::preExecute();
```

Comment:
so preExecute is not really hard-coded to "TIMESTEP_BEGIN" in that base class. Seems odd to hard-code it in this class as well. 
I could see two remedies:
- check the execute_on here and assert it to be that timestep_begin
- make a more general "preExecute ( execute_on) "

Reply:
agreed.
However DefaultMultiAppFixedPointConvergence::preExecute() should not be called outside of TIMESTEP_BEGIN because of the hardcoded log